### ISSUE TITLE
Allow for exceptions on `disallowAnonymousFunctions`

### ISSUE THREAD

**Issue by twolfson** [jscs-dev/node-jscs#834]
I prefer to name my anonymous functions most of the time (e.g. in a `forEach`) for better stack traces. However, sometimes it is unnecessary, redundant, and impractical (e.g. inside of each `it` declaration in my tests):

```js
describe('An apple', function () {
  it('is red', function assertRed () {
    // Redundant name and 2 items to maintain
  });
});
```

I would like to allow for exceptions on `disallowAnonymousFunctions`. This can get tricky since anonymous functions can occur in assignment, as a parameter, and probably some other cases I can't think of at the moment.

For now, I am proposing adding `parameterExcept` which allows for exceptions of anonymous functions that are used inside of a `CallExpression` with the final `Identifier` matching a value.

```js
"disallowAnonymousFunctions": {
  "parameterExcept": ["it"]
}
```

**Comment by markelog** [comment]
Sorry for the late response, got caught up with other stuff. Your use-cases sound like a new rule, if you send us a PR we would definitely consider it.

It seems verbose error stacks is primary reason for use of `disallowAnonymousFunctions` i would extend its scope. 

However, function might be a named expression, which might be considered as an error, if this rule would find these cases. But i would catch them anyway.

@mikesherov what's your take on that?

**Comment by markelog** [comment]
Relevant https://github.com/jscs-dev/node-jscs/issues/834
/cc @Nijikokun, @ahmadnassri

**Comment by Nijikokun** [comment]
Relevant to #655

**Comment by markelog** [comment]
oops, wrong reference, thanks @Nijikokun

**Comment by twolfson** [comment]
Yep, it looks like @Nijikokun and I are on the same page about the rule we want. I want to close this issue as a duplicate but it looks like the thread has shifted over here instead of stayed on #655 =/

**Comment by markelog** [comment]
> duplicate but it looks like the thread has shifted over here instead of stayed on #655 =/

As long as it would be fixed, i don't care where is discussion is held :). Is there any takers for this one?We are pretty buried with other issues.

**Comment by twolfson** [comment]
I am down for submitting a PR. However, I would really like the exceptions to the rule we are creating (e.g. `it`). I am having trouble nailing down a good API signature though; it should support exceptions for:

```js
it(function () {
it.skip(function () {
```

**Comment by twolfson** [comment]
I will probably call the rule `requireStackNamedFunctions` since that is our use case.

**Comment by twolfson** [comment]
I think I am going to solve this by ignoring dynamic key identifiers for now. I think that is an edge case that we will have to deal with when we get there.

```js
"requireStackNamedFunctions": true

"requireStackNamedFunctions": {
  "allExcept": ["it", "it.skip"]
}
```

**Comment by twolfson** [comment]
Choosing a different name, `requireNamedInlineFunctions`, since the negation of `StackNamedFunctions` would prevent usage of any left hand assignment which is far from the intent.

**Comment by mikesherov** [comment]
I agree with @markelog here.

**Comment by twolfson** [comment]
I have created a PR for to fix this issue and #655. It is #891.

I decided to rename the rule once more to `requireNamedUnassignedFunctions` since the BDD descriptions were getting confusing when negating.

**Comment by zxqfox** [comment]
@twolfson Thanks for your patience and contribution. Well done!
